duffy/app/controllers/node.py

```diff
@@ -22,7 +22,7 @@
 @router.get("", response_model=NodeResultCollection, tags=["nodes"])
 async def get_all_nodes(
     db_async_session: AsyncSession = Depends(req_db_async_session),
-    tenant: Tenant = Depends(req_tenant),
+    tenant: Tenant = Depends(req_tenant_optional),
```

Comment:
Hmm… previously, just having the dependency ensured that authenticated tenants would see the list of nodes. Now we want to be able to cater for the semantics of the legacy API in the metaclient, i.e. return all (active) nodes if not executed in the context of a specific tenant there. I don't think we should do it like this here, though (it's a terrible side effect IMO), but we could do it like this:

- Add an optional query parameter e.g. `filter_for_tenant`. If unset, it would return only nodes "assigned" to an authenticated normal tenant, or return all active nodes for admins or if unauthenticated. This will surely make querying more complicated, i.e. we should probably select on active `Session` objects instead and pull the list of nodes from there. This would give us more code paths to cover in the associated tests 🙈😉!
- Implement filtering for only active nodes (this is an oversight so far, we may want this in other endpoints/functions, too if we don't have it yet). It would need an optional query parameter (e.g. `only_active` which defaults to `True`). Admins (and only them) could use this if they really want know about retired nodes.

Reply:
@nphilipp, we need not complicate our base application to play well with the design measures taken by the legacy Duffy that might not be relevant anymore. I have added a dependency that plays along with both the presence and absence of the credentials that you can find here https://github.com/CentOS/duffy/pull/184/files#diff-72c56f63f831ba4972b4c0a40879fcacff391572dfc8e2bcd798bd29b0ac3580R17-R25.

Let me elaborate upon how it works.

1. For the path operations that **require authentication** (irrelevant of the user being client or admin as that is something taken care of by the session created by the new Duffy when the non-empty credentials are passed), they would work just fine in both **allowing** (should the credentials _be correct_ AND _be present in our config file_ for old users) and **disallowing** (should either the credentials _be incorrect_ OR their credentials _be absent in our config file_ for old users).

2. For the path operations that **do not necessitate authentication** (irrelevant of the credentials being passed or not - let alone the case when those are actually passed but still irrelevant of the user being client or admin as that is something taken care of by the session created by the new Duffy when the non-empty credentials are passed), they would work just fine in both **allowing to show ALL nodes** (should the credentials _not be passed_ OR should the credentials _be passed_, and the users is an _admin_ with their credentials _being correct_ and _present in our config file_ for old users) and **allowing to show PER-USER nodes** (should the credentials _be passed_, _be correct_ AND _be present in our config file_ for old users).

Reply:
IMO what I described is not just to cater for legacy idiosyncracies, but real fixes for the API:

- Allow unauthenticated querying of nodes which we didn't before. In theory we could only allow that in the legacy API, but that's an odd picture: you get easier access there than in the current API which we want people to move to. Also, we'd need some trickery so the metaclient can query the nodes in that case because we can't use the tenant's credentials.
- Only return nodes which are relevant to a tenant by default – this is what they're presumably interested in. For the sake of simplicity, let's make it: normal tenants get nodes which are currently assigned to them, admin and unauthenticated get all nodes. Does that sound sane?
- Only return active nodes which we should have done from the start. Querying retired nodes is out of scope for this, too.

Reply:
This one's still open :wink:, if we don't filter for tenant, we don't need the `req_tenant_optional` dependency at all (we had `req_tenant` only for the side effect of ensuring a logged in tenant).